Subdecoders and json schema

azahnen opened the pull request:
<!--
    Thanks for taking the effort to create a PR! 🙌

    👋 Are you making a change to documentation only? Delete the rest of the template and go ahead.

    👋 If you are creating an extensive PR, you might want to open an issue with your idea first in case there is a chance for rejecting it.

    👋 If you intend to work on a PR over several days, please create [draft pull requests](https://github.blog/2019-02-14-introducing-draft-pull-requests/) instead.

    👇 Please look at the following checklist to ensure that your PR can be accepted quickly (feel free to ask for help):
-->

### Pull request checklist

-   [ ] Added/updated unit tests
-   [x] Updated documentation
-   [x] All checks are passing


### Changes introduced by this PR

<!--
    Either reference the issue(s) that describe the changes introduced by this PR 
-->

Closes #907.
Closes #908.
Closes #931.

Depends on interactive-instruments/xtraplatform-spatial#195.

<!--
    or shortly describe the changes below using bullet points
-->

azahnen commented:
<!--woodpeckerci-plugin-surge-preview-->
Deploying preview to https://interactive-instruments-ldproxy-pr-944.surge.sh

cportele commented:
@azahnen 

I have problems to get started with testing. Here is a DB table that I am using:

```sql
CREATE TABLE public.features (
	"_id" serial4 NOT NULL,
	id text NULL,
	properties jsonb NULL,
	geom public.geometry NULL,
	fid int4 NULL,
	featuretype text NULL,
	CONSTRAINT features_pkey PRIMARY KEY (_id)
);
```

I tried `auto` mode, which was not successful: the `jsonb` column was ignored and there was an exception that the FeatureSchema could not be created because `name` was missing. However, a schema was created, but an incomplete one.

I have then updated to schema as follows:

```yaml
types:
  features:
    sourcePath: /features
    type: OBJECT
    allOf:
    - properties:
        _id:
          sourcePath: _id
          type: INTEGER
          role: ID
        id:
          sourcePath: id
          type: STRING
        featuretype:
          sourcePath: featuretype
          type: STRING
        fid:
          sourcePath: fid
          type: INTEGER
        geom:
          sourcePath: geom
          type: GEOMETRY
          role: PRIMARY_GEOMETRY
          geometryType: ANY
    - sourcePath: '[JSON]properties'
      schema: https://www.portele.de/xplanung/pfs_properties.json
```

I can access the features, but `properties` is ignored. I did have the time to debug the reasons. Do you have an idea what I am doing wrong? Thanks!

I used log level DEBUG, but there were no messages related to problems in the log.

azahnen commented:
@cportele I started to implement unit tests for the schema resolver and looked into https://www.portele.de/xplanung/pfs_properties.json to see if I could add that. The top-level `oneOf` is problematic since there is no way to select an element, it will always take the first one. If the `oneOf` would consist of `$ref` elements, you could reference the according `$defs` in `schema`.

Do you still work with that schema or do you have a customized variant somewhere?

azahnen commented:
@cportele The `oneOf` would work the same but would not play a role anymore since you can directly reference `$defs`. Since I assumed that there is one `oneOf` element per feature type, you could then do something like this: 
```yaml
schema: https://www.portele.de/xplanung/pfs_properties.json#/$defs/PFS_Schutzgebiet
```

azahnen commented:
@cportele I was able to reproduce the issue with nested objects (`Invalid sourcePath in provider configuration`), I pushed a fix.

cportele commented:
@azahnen - Yes, I am testing it now, this seems to fix this exception. Thanks.

cportele commented:
@azahnen - I have created a `jsontest` repo with the configuration that I am using for testing.

Here are some issues that I have noticed so far: 

- [ ] For the first feature (in each request) the properties are as expected. Starting with the second feature, the properties from the JSON column are embedded in an `_IGNORE_` member.
- [ ] `SchemaDeriverOpenApi.withName()` throws an exception, because `oapiSchema` is `null` (the schema is identified as a value, but is not a value and therefore stays `null`).
- [ ] The order of the properties does not follow the order in the JSON Schema. I guess that is not really an issue, more an observation since we probably cannot control how a) the properties in the `properties` member are parsed when reading the schema and b) for a `jsonb` column the order of members in a object is not predictable. Or is there a way to process the schema so that the order in the feature schema is the same as in the JSON Schema? Then using a `json` column should maintain the order?

Could you have a look at these issues?

azahnen commented:
@cportele 
- The first two issues should be fixed. 
- I could not reproduce the third issue, even without the recent fixes. How can I trigger the exception?
- Regarding the sort order, order from `json` columns should be maintained now. The sorting transformer falsely reordered some json properties, it should ignore them for now. 
I also looked into making the sorting transformer more generic so that it could reorder the properties for `jsonb` columns, but that would be quite complex. Currently it only handles a very specific case where values from a row are intercepted by a joined query. Handling any order and missing properties would require a completely different mechanism.
Note that all of this has nothing to do with the schema, the feature schema retains the order from the JSON Schema.

cportele commented:
@azahnen 

Yes, the first two issues are resolved. 

From the description it also sounds like we should consider the ordering issue resolved - I will do some tests over the weekend with `json` columns. This is mainly relevant, if there is a format where the order is important (GML).

I also no longer see the exception. It was still there yesterday (a type of a property was `UNKNOWN`), but maybe that was before the changes to the JSON Schema. I will also do more tests with schema variations.

Thanks!

cportele commented:
@azahnen - I have made more tests and I had to make some code changes to get the correct/better results. Since some of that may not be the right or best way, I have done this in separate branches. Change overview:

ldproxy (`subdecoders-and-json-schema-2`):

* remove `_IGNORE_` in OpenAPI Schemas
* remove `_IGNORE_` in JSON Schemas of returnables
* remove `_IGNORE_` in the values of the `properties` parameter

xtraplatform-spatial (`subdecoders-2`):

* `[JSON]column_name` did not work for a property. This was mapped to a column called `JSON`.
* JSON values that are, e.g., just a number were not parsed, the token was `NOT_AVAILABLE`. The proper approach as I understand it would be to send `endOfInput()` to the feeder, but this conflicts with the way the parser is used. For now I have added a whitespace character at the end so that the parser can provide the correct token, but that is a hack.
* I have added warnings when `allOf`/`anyOf`/`oneOf` are encountered as well as code to process constraints that can be mapped to `SchemaConstraints`.

There several TODOs and unfinished new code (e.g. `GEO_DEFS` only has polygon geometries). Before merging into master, we should clean-up the new code.

In addition, we should add more documentation about the expectations/limitations of `schema` in https://interactive-instruments-ldproxy-pr-944.surge.sh/de/providers/feature/#types.

cportele commented:
@azahnen 

I did tests and found more issues: 

* `_IGNORE_` is also included in the property names after flattening, e.g., in CSV. This would require changes to `WithTransformationsApplied()`. Instead of adapting more existing (and future) code with treatment for this special case it would be better to find another approach so that we can avoid code changes that will be difficult to maintain.
* Flattening values from JSON columns also seems to require special treatment. A JSON value `[ 1.2, 2.4 ]` of a property `test` (`type: VALUE_ARRAY`, `valueType: FLOAT`) is flattened to two properties `test.test`, which in turn results in an map key exception because there are two properties/keys with the same name.

azahnen commented:
@cportele

**json colums**
- I added support for values and value arrays.

**allOf**
- I replaced the `_IGNORE_` logic with a better aproach as discussed.

These two changes should fix all issues regarding flattening and schemas.

**external schema**
- I added support for reusable schema `fragments`. By default only entries from `fragments` can be referenced in `schema`.
- I moved the JSON schema resolver to an extension that has to be enabled to use JSON schemas in `schema`. That makes it easier to add configuration options and also adds a separate page in the documentation.
- I added a configuration option to map `$defs` to `objectType`s, so you do not have to define all `Link`s manually.
- I merged the constraints derivation from your branch.
- I did not merge the warnings. I think warnings that the user cannot fix are just annoying. It should be enough that the limitations are described in the documentation.
- I updated the `jsontest` repo to make use of the improvements.

cportele commented:
@azahnen - I checked all issues and they are all gone now. I will do more testing tomorrow. 

`fragments` is helpful and to make `JSON_SCHEMA` an extension looks like the right approach to me.

> I did not merge the warnings. I think warnings that the user cannot fix are just annoying. It should be enough that the limitations are described in the documentation.

This is something that we should discuss. I am more annoyed by the silent re-interpretation of the schema (and I did adapt/fix the schema/configuration, once I found out that the oneOf was ignored).

azahnen commented:
@cportele I could imagine logging the warnings when either `typeValidation` or a new flag in `JSON_SCHEMA` is enabled. But there should be some way to get rid of the warnings besides setting the log level to `ERROR`.

cportele commented:
@azahnen - Moving the warning checks to `typeValidation` during start-up would probably be best. I will have a look.

azahnen commented:
@cportele They can stay in the same place in `JsonSchemaResolver.resolveComposition`. You just have to pass `data` as parameter and check for `typeValidation != NONE`.
